aws/resource_aws_guardduty_publishing_destination.go

```diff
@@ -0,0 +1,205 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/guardduty"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+// Constants not currently provided by the AWS Go SDK
+const (
+	guardDutyPublishingStatusFailed = "FAILED"
+)
+
+func resourceAwsGuardDutyPublishingDestination() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsGuardDutyPublishingDestinationCreate,
+		Read:   resourceAwsGuardDutyPublishingDestinationRead,
+		Update: resourceAwsGuardDutyPublishingDestinationUpdate,
+		Delete: resourceAwsGuardDutyPublishingDestinationDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"detector_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"destination_type": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  guardduty.DestinationTypeS3,
+			},
+			"destination_arn": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validateArn,
+			},
+			"kms_key_arn": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validateArn,
+			},
+		},
+	}
+}
+
+func resourceAwsGuardDutyPublishingDestinationCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).guarddutyconn
+
+	detectorID := d.Get("detector_id").(string)
+	input := guardduty.CreatePublishingDestinationInput{
+		DetectorId: aws.String(detectorID),
+		DestinationProperties: &guardduty.DestinationProperties{
+			DestinationArn: aws.String(d.Get("destination_arn").(string)),
+			KmsKeyArn:      aws.String(d.Get("kms_key_arn").(string)),
+		},
+		DestinationType: aws.String(d.Get("destination_type").(string)),
+	}
+
+	log.Printf("[DEBUG] Creating GuardDuty publishing destination: %s", input)
+	output, err := conn.CreatePublishingDestination(&input)
+	if err != nil {
+		return fmt.Errorf("Creating GuardDuty publishing destination failed: %s", err.Error())
+	}
+
+	stateConf := &resource.StateChangeConf{
```

Comment:
Could you please move this code to `aws/internal/service/guardduty/waiter` and follow the pattern there?
See https://github.com/terraform-providers/terraform-provider-aws/issues/12840.